### DISCUSSION TITLE
Receiving Data from COM PORT devices

### DISCUSSION THREAD

**Discussion by wungfuh** [collin80/SavvyCAN#461] [General]
Does SaavyCAN interact automatically with devices connected via COM port? I have a custom device that was designed by me lecturer to serve as a CAN Bus sniffer. It has an STM32F7 as MCU and I want to program it to serve as a Communication Device class to emulate the serial port on the PC. I wish to connect the device directly to the PC without a CAN-USB adapter.
Please be basic with me, I am a newbie trying to chew hard meat.

**Comment by cburlacu** [comment]
Your device must "speak" one of the protocols supported by SavvyCAN: [GVRET](https://github.com/collin80/GVRET), slcan, etc

From all of those `slcan` is the easiest way to implement: https://github.com/torvalds/linux/blob/master/drivers/net/can/slcan.c:
``` 
* Examples:
 *
 * t1230 : can_id 0x123, len 0, no data
 * t4563112233 : can_id 0x456, len 3, data 0x11 0x22 0x33
 * T12ABCDEF2AA55 : extended can_id 0x12ABCDEF, len 2, data 0xAA 0x55
 * r1230 : can_id 0x123, len 0, no data, remote transmission request
 ```
 Basically you need to convert each CAN message to a `slcan` and write it on USB CDC "serial port"...

**Reply by wungfuh** [reply]
I defined my STM32F7 board as a WINUSB device, when I go to connection type menu of SavvyCan, I find only serial connection and the default protocols is GVRET. How do I connect my WINUSB device  as USB and using the slcan protocol formate
![image](https://user-images.githubusercontent.com/60786353/163546556-44f35db3-f6fc-402d-b0d8-2af3a6188dc2.png)

**Reply by collin80** [reply]
Serial connection in this case means the GVRET protocol. It's a custom thing I made like a decade ago and then got stuck with. For SLCAN you'd need to be using LINUX and then the slcan socketcan tools. I am thinking of adding direct support for lawicel protocol which would allow SLCAN devices in Windows as well.

**Reply by wungfuh** [reply]
@collin80 Thanks for reply. What is the data format of the GVRET protocol? Can I send raw CAN data or do I need to frame it to the GVRET protocol?

**Reply by cburlacu** [reply]
You may need to port this file https://github.com/collin80/GVRET/blob/master/GVRET.ino into your project. It seems that is the file responsible converting a CAN frame to _GVRET protocol_

**Reply by amiralnar** [reply]
@collin80 Hi! Thank you for developing this great program. I like it.
Is there any way to help you implement direct support for lawicel protocol?
I'm newbie in QT programming but i want it help and i can do some stuff.
